compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -944,6 +944,47 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
         interp_ok(())
     }
 
+    pub fn prepare_for_native_call(
+        &mut self,
+        id: AllocId,
+        initial_prov: M::Provenance,
+    ) -> InterpResult<'tcx> {
+        // Expose provenance of the root allocation.
+        M::expose_provenance(self, initial_prov)?; // TODO: Is this the right way to expose provenance?
+
+        let mut done = rustc_data_structures::fx::FxHashSet::default();
+        let mut todo = vec![id];
+        while let Some(id) = todo.pop() {
+            if done.insert(id) {
+                // This is a new allocation, add the allocation it points to `todo`.
+                let info = self.get_alloc_info(id);
+
+                // If there is no data behind this pointer, skip this.
+                if !matches!(info.kind, AllocKind::LiveData) {
+                    continue;
+                }
+
+                let alloc = self.get_alloc_raw(id)?;
+                for prov in alloc.provenance().provenances() {
+                    //M::expose_provenance(self, prov)?; // TODO: Is this the right way to expose provenance? + mutable borrow here gives issues due to provenances iterator lifetime...
```

Comment:
The best option is probably to change the signature of `expose_provenance` to take `&self` (and then fix everything in Miri to make it build again)

Reply:
I tried this now and so far it seems to work - mostly had to change `let this = self.eval_context_mut();` to `let this = self.eval_context_ref();` in a few places.